### PR TITLE
Implement Activity.AddLink

### LINKED ISSUE #97680: Support adding links to Activity after it is created
OpenTelemetry has added [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#link) for supporting link addition to spans after the span is created. PR marking spec as stable - https://github.com/open-telemetry/opentelemetry-specification/pull/3887.

 This is currently not possible in .NET. Links can only be added when the activity is being created see [this](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activitysource.createactivity?view=net-8.0#system-diagnostics-activitysource-createactivity(system-string-system-diagnostics-activitykind-system-string-system-collections-generic-ienumerable((system-collections-generic-keyvaluepair((system-string-system-object))))-system-collections-generic-ienumerable((system-diagnostics-activitylink))-system-diagnostics-activityidformat))

Opening the issue for tracking as supporting this new feature would require addition of new API in System.Diagnostics.DiagnosticSource.

### API Proposal

```C#
namespace System.Diagnostics
{
    public partial class Activity : IDisposable
    {
        public System.Diagnostics.Activity AddLink(System.Diagnostics.ActivityLink link);
    }
}
```

### PR DESCRIPTION
**Pull request by antonfirsov**
Resolves #97680 assuming the API will be approved, marking as NO-MERGE until then.

### CODE REVIEW

**Comment by stephentoub** [comment]
> assuming the API will be approved

It's not marked as api ready for review, or even api suggestion... it's not going to show up on anyone's api approval radar.

**Comment by tarekgh** [comment]
I am closing this till we review and approve the API. We can reactivate after then.

**Comment by tarekgh** [comment]
I marked https://github.com/dotnet/runtime/issues/97680 as ready for review.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs [hunk 1]
```diff
@@ -1597,6 +1597,31 @@ public void TestEvent()
...
+
+            ActivityLink l1 = new ActivityLink(c1);
+            ActivityLink l2 = new ActivityLink(c2, new ActivityTagsCollection()
+            {
+                new KeyValuePair<string, object?>("foo", 99)
+            });
+
+            Activity activity = new Activity("LinkTest");
+            Assert.True(ReferenceEquals(activity, activity.AddLink(l1)));
+            Assert.True(ReferenceEquals(activity, activity.AddLink(l2)));
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs:1614]: Do we need a test that validates what happens (which presumably means "doesn't blow up") if the same ActivityLink is added twice?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs:1614]: @antonfirsov Also, could you please add some more tests inside the test method [EnumerateLinksTest()](https://github.com/dotnet/runtime/blob/40bc2d88b6606324f5774cc972d480a1d26084f8/src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs#L2143)?

**Comment by tarekgh** [comment]
CC @noahfalk @samsp-msft @reyang @cijothomas @CodeBlanch @vishweshbankwar

**Review by noahfalk — APPROVED** [review]
LGTM

**Review by lmolkova — APPROVED** [review]

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs (hunk 1 above)
Comment (antonfirsov) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs:1614]: Fixed in [b367037](https://github.com/dotnet/runtime/pull/101381/commits/b367037c92a916bab076127e8fad3c1df78a3d0e).

**Comment by tarekgh** [comment]
@antonfirsov could you please merge the main branch to yours to proceed with this PR? Thanks!

**Pull request merged**